plutus-ir/src/Language/PlutusIR/Rename.hs

```diff
@@ -0,0 +1,131 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE LambdaCase       #-}
+{-# OPTIONS_GHC -fno-warn-orphans #-}
+-- | Renaming of PIR terms. Import this module to bring the @PLC.Rename (Term tyname name ann)@
+-- instance in scope.
+module Language.PlutusIR.Rename () where
+
+import           Language.PlutusIR
+
+import qualified Language.PlutusCore                 as PLC
+import qualified Language.PlutusCore.Name            as PLC
+import qualified Language.PlutusCore.Rename          as PLC
+import qualified Language.PlutusCore.Rename.Internal as PLC
+
+import           Control.Monad.Reader
+import           Control.Monad.Trans.Class           (lift)
+import           Control.Monad.Trans.Cont            (ContT (..))
+
+{- Note [Renaming of mutually recursive bindings]
+The 'RenameM' monad is a newtype wrapper around @ReaderT renaming Quote@, so in order to bring
+a name into the scope we need to use the following pattern:
+
+    withFreshenedX :: X -> (X -> RenameM renaming c) -> RenameM renaming c
+
+where an 'X' is some value that has names inside and the second argument is a continuation that
+expects to receive this value, but with all the names freshened and with the mappings from the
+old uniques to the new ones saved. For the latter reason we can't just write
+
+    freshenX :: X -> RenameM renaming X
+
+because if you write
+
+    freshenX x >>= \xFr -> ...
+
+then the mappings from the uniques of 'x' to the uniques of 'xFr' do not exist in @...@.
+
+Two problems arise:
+
+1. If you have a list of @X@s and then an @Y@ and then another list of @Z@s and you need to freshen
+   all of those, it's not nice to compose @withFreshened*@ functions at all. In order to make these
+   functions composable, we can wrap them in the 'ContT' monad. See e.g. 'renameDatatypeCM' below.
+2. Handling of mutually recursive bindings requires to bring a bunch of bindings in scope first and
+   only then to rename their RHSs. I.e. one part of a 'ScopedRenameM' computation must be performed
+   immediately (brinding names in scope) and the other part must be postponed until all bindings
+   are in scope (renaming of RHSs). Hence we have the following pattern:
+
+       renameXCM :: X -> ContT c PLC.ScopedRenameM (PLC.ScopedRenameM X)
+
+   where an 'X' is some value that defines some names and has types/terms that need to be renamed.
+   The first 'PLC.ScopedRenameM' is for bringing names in scope and the second 'PLC.ScopedRenameM'
+   is for performing the renaming.
```

Comment:
Hm, I'm vacillating on this. I'm not sure I particularly like either of the ways, it seems like there should be a better way.

Reply:
> Then you can also stick all the other functions in `Cont` and it works quite nicely. It's a bit annoying that `withFresehenedVarDecl` is less useful and you just have to rename the bindings like variables on the second pass, but I think it's okay.

I definitely like the one-pass approach more, because it's kinda more encapsulated while with the two-pass approach you need to connect the first pass with the second one via some implicit "knowledge" that it's sensible to do so. The "knowledge" seems like an invariant, but I do not know how to express it to make it less implicit.

The one-pass approach does not require to rename the names of bindings (only to freshen them) which is nice. Do you have some code I can look at?